src/App.tsx

```diff
@@ -1,61 +1,153 @@
+  const [todos, setTodos] = useState<TodoWithUser[]>(preparedTodos);
+  const [title, setTitle] = useState('');
+  const [userId, setUserId] = useState(0);
+  const [isTitleError, setIsTitleError] = useState(false);
+  const [isUserIdError, setIsUserIdError] = useState(false);
+
+  const handleTitleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setTitle(event.target.value);
+    setIsTitleError(false);
+  };
+
+  const handleAddTodo = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    let isError = false;
+
+    if (!title) {
+      setIsTitleError(true);
+      isError = true;
+    }
+
+    if (userId === 0) {
+      setIsUserIdError(true);
+      isError = true;
+    }
+
+    if (isError) {
+      return;
+    }
+
+    function resetForm() {
+      setTitle('');
+      setUserId(0);
+    }
+
+    const todoIds = todos.map(({ id }) => id);
+    const maxTodoId = Math.max(...todoIds);
+
+    const newTodo: TodoWithUser = {
+      id: maxTodoId + 1,
+      title,
+      completed: false,
+      userId,
+      user: findUserById(userId),
+    };
+
+    setTodos([...todos, newTodo]);
+
+    resetForm();
+  };
+
+  const handleSelectUserId = (event: React.ChangeEvent<HTMLSelectElement>) => {
+    setUserId(+event.target.value);
+
+    setIsUserIdError(false);
+  };
+
   return (
     <div className="App">
       <h1>Add todo form</h1>
 
-      <form action="/api/todos" method="POST">
+      <form
+        action="/api/todos"
+        method="POST"
+        noValidate
+        onSubmit={handleAddTodo}
+      >
         <div className="field">
-          <input type="text" data-cy="titleInput" />
-          <span className="error">Please enter a title</span>
+          <label
+            htmlFor="title"
+          >
+            Title:
+          </label>
+
+          <input
+            id="title"
+            type="text"
+            data-cy="titleInput"
+            value={title}
+            onChange={handleTitleChange}
+            required
+            placeholder="Enter a title"
+          />
+          {isTitleError && (
+            <span className="error">Please enter a title</span>)}
         </div>
 
         <div className="field">
-          <select data-cy="userSelect">
-            <option value="0" disabled>Choose a user</option>
+          <label
+            htmlFor="user-id"
+          >
+            User:
+          </label>
+
+          <select
+            id="user-id"
+            data-cy="userSelect"
+            value={userId}
+            onChange={handleSelectUserId}
+            required
+          >
+            <option
+              value="0"
+              disabled
+            >
+              Choose a user
+            </option>
+            {usersFromServer.map((user: User) => (
+              <option
+                key={`${user.name}_user.name`}
+                value={user.id}
+              >
+                {user.name}
+              </option>
+            ))}
```

Comment:
It can be written shorter:
```suggestion
            {usersFromServer.map(({ id, name }) => (
              <option key={id} value={id}>
                {name}
              </option>
            ))}
```